cadence/server/api.go

```diff
@@ -0,0 +1,252 @@
+// handlers.go
+// API functions and fileservers
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/dhowden/tag"
+	"github.com/kenellorando/clog"
+)
+
+// /api/search
+// Gets text metadata (excludes album art and path) of any songs matching a search query.
+func Search() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		clog.Info("Search", fmt.Sprintf("Search request from client %s.", r.Header.Get("X-Forwarded-For")))
+
+		type Search struct {
+			Query string `json:"search"`
+		}
+		var search Search
+
+		decoder := json.NewDecoder(r.Body)
+		err := decoder.Decode(&search)
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest) // 400 Bad Request
+			return
+		}
+
+		queryResults, err := searchByQuery(search.Query)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+
+		jsonMarshal, _ := json.Marshal(queryResults)
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(jsonMarshal)
+	}
+}
+
+// /api/request/id
+// Posts a request submission for a specific song ID.
+func RequestID() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		clog.Info("Request", fmt.Sprintf("Request-by-ID by client %s.", r.Header.Get("X-Forwarded-For")))
+
+		type Request struct {
+			ID string `json:"ID"`
+		}
+		var request Request
+
+		decoder := json.NewDecoder(r.Body)
+		err := decoder.Decode(&request)
+		if err != nil {
+			clog.Error("RequestID", "Unable to decode request.", err)
+			w.WriteHeader(http.StatusBadRequest) // 400 Bad Request
+			return
+		}
+		reqID, err := strconv.Atoi(request.ID)
+		if err != nil {
+			clog.Error("RequestID", "Unable to convert request ID to an integer.", err)
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		path, err := getPathById(reqID)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+		_, err = liquidsoapRequest(path)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+
+		w.WriteHeader(http.StatusAccepted) // 202 Accepted
+	}
+}
+
+// /api/request/bestmatch
+// Posts a request submission for the top result of a search.
```

Comment:
Hey, this is familiar!!

Wow, this might actually be the first time I ever trim *down* CadenceBot, but I like having one-step-request not happen inside my code...

Reply:
I thought you would like this. Directly inspired by CadenceBot.